File: server/test/autoboard/comments_test.exs
defmodule Autoboard.CommentsTest do
  use Autoboard.DataCase, async: false

  import Ecto.Query

  alias Autoboard.Activity.Event
  alias Autoboard.Auth.Context
  alias Autoboard.Comments
  alias Autoboard.Domain.Error
  alias Autoboard.Projects
  alias Autoboard.Repo
  alias Autoboard.Tickets

  setup do
    %{ctx: Context.global(:codex)}
  end

  test "adds an append-only comment using the authenticated actor and increments revision", %{
    ctx: ctx
  } do
    project = project_fixture(ctx, "AUTO")
    ticket = ticket_fixture(ctx, project)

    assert {:ok, comment} = Comments.add(ctx, ticket.id, %{body: "A useful note"})
    assert comment.actor == :codex
    assert comment.body == "A useful note"
    assert comment.ticket_id == ticket.id
    assert comment.project_id == project.id

    assert {:ok, updated_ticket} = Tickets.fetch(ctx, ticket.id)
    assert updated_ticket.revision == ticket.revision + 1

    assert %Event{event_type: "comment.added", actor: :codex, ticket_id: ticket_id} =
             Repo.one(
               from(event in Event,
                 where: event.ticket_id == ^ticket.id,
                 order_by: [desc: event.id],
                 limit: 1
               )
             )

    assert ticket_id == ticket.id
  end

  test "rejects blank comments without changing the ticket or activity log", %{ctx: ctx} do
    project = project_fixture(ctx, "AUTO")
    ticket = ticket_fixture(ctx, project)

    assert {:error, %Error{kind: :validation_failed, fields: %{body: [_]}}} =
             Comments.add(ctx, ticket.id, %{body: " \n\t "})

    assert {:ok, unchanged} = Tickets.fetch(ctx, ticket.id)
    assert unchanged.revision == ticket.revision
    assert Repo.aggregate(Event, :count) == 2
  end

  test "preserves public authorization and archived project guards", %{ctx: ctx} do
    project = project_fixture(ctx, "AUTO")
    ticket = ticket_fixture(ctx, project)

    assert {:error, %Error{kind: :unauthorized}} =
             Comments.add(%Context{actor: :system, scope: :global}, ticket.id, %{body: "Nope"})

    assert {:ok, archived} = Projects.archive(ctx, project.id, project.revision)

    assert {:error, %Error{kind: :validation_failed}} =
             Comments.add(ctx, ticket.id, %{body: "Nope"})

    assert archived.state == :archived
  end

  defp project_fixture(ctx, key) do
    assert {:ok, project} = Projects.create(ctx, %{key: key, name: "Project #{key}"})
    project
  end

  defp ticket_fixture(ctx, project) do
    assert {:ok, ticket} = Tickets.create(ctx, %{project_id: project.id, title: "Ticket"})
    ticket
  end
end

File: internal/app/comments_test.go
package app_test

import (
	"context"
	"errors"
	"testing"

	"github.com/m-cain/autoboard/internal/app"
)

func TestAddCommentAppendsAndAdvancesTicketRevision(t *testing.T) {
	service := openService(t)
	ctx := context.Background()
	project, err := service.CreateProject(ctx, app.CreateProjectInput{Key: "AUTO", Name: "Auto"})
	if err != nil {
		t.Fatalf("create project: %v", err)
	}
	ticket := createTicket(t, service, project.ID, "Commented")

	comment, ticket, err := service.AddComment(ctx, ticket.ID, "A useful note")
	if err != nil {
		t.Fatalf("add comment: %v", err)
	}
	if comment.Body != "A useful note" ||
		comment.Actor != app.ActorCodex ||
		comment.TicketID != ticket.ID ||
		comment.ProjectID != project.ID {
		t.Errorf("comment = %#v", comment)
	}
	if ticket.Revision != 2 {
		t.Errorf("ticket revision = %d, want 2", ticket.Revision)
	}
	events, err := service.ListActivityAfter(ctx, 0)
	if err != nil {
		t.Fatalf("list activity: %v", err)
	}
	if got := events[len(events)-1].EventType; got != "comment.added" {
		t.Errorf("last event = %q, want comment.added", got)
	}
}

func TestAddCommentRejectsBlankAndArchivedProject(t *testing.T) {
	service := openService(t)
	ctx := context.Background()
	project, err := service.CreateProject(ctx, app.CreateProjectInput{Key: "AUTO", Name: "Auto"})
	if err != nil {
		t.Fatalf("create project: %v", err)
	}
	ticket := createTicket(t, service, project.ID, "Commented")

	_, _, err = service.AddComment(ctx, ticket.ID, " \n\t ")
	var domainErr *app.Error
	if !errors.As(err, &domainErr) || domainErr.Kind != app.ErrorValidationFailed {
		t.Fatalf("blank comment error = %v, want validation_failed", err)
	}
	_, err = service.ArchiveProject(ctx, project.ID, project.Revision)
	if err != nil {
		t.Fatalf("archive project: %v", err)
	}
	_, _, err = service.AddComment(ctx, ticket.ID, "Nope")
	if !errors.As(err, &domainErr) || domainErr.Kind != app.ErrorInvalidTransition {
		t.Fatalf("archived comment error = %v, want invalid_transition", err)
	}
	unchanged, err := service.GetTicket(ctx, ticket.ID)
	if err != nil {
		t.Fatalf("reload ticket: %v", err)
	}
	if unchanged.Revision != ticket.Revision {
		t.Errorf("ticket revision = %d, want %d", unchanged.Revision, ticket.Revision)
	}
}
